### DISCUSSION TITLE
GitHub Universe 2020 Day 2: What did you learn?

### DISCUSSION THREAD

**Discussion by siddharthdayalwal** [EddieHubCommunity/support#537] [Conference Notes]
From the [GitHub Universe 2020](http://githubuniverse.com/) please share what you learnt or any tools you want to learn more about below :point_down: 
![GitHub Univers](https://user-images.githubusercontent.com/624760/101410113-c5587e80-38d6-11eb-9d34-fe8e20f70a55.png)

**Comment by siddharthdayalwal** [comment]
Will be sharing notes from **[GitHub Univers(ity) 2020 - Day 2](https://www.twitch.tv/githubeducation)**  🎉

# **`The university lab: experiment with the pack`** :package:  
> Get all [GitHub Education Benefits](https://education.github.com/)
- - -

## [Applying the game design stack](https://www.twitch.tv/videos/831625368) :video_game:  
**Kevin Whinnery - Senior Manager, TwilioQuest, Twilio**
> The “game design stack” is a concept used by developers to describe the experience of playing a game. Fantasy, action, economy, world, and story come together to form the DNA of a game, and provide a framework for making decisions about how your game works. Finding out which of these design elements is most important will help you make tough decisions about where to invest your time and energy. In this session, Kevin will begin by describing the game design stack at a high level, and how it can be a helpful tool for articulating what your game is about, and making decisions on where to invest your time. He’ll then explore the practical implications with an example from building TwilioQuest, an educational PC game for developers, and introduce you to the toolchain you need to build games with web development skills.

Gist of the talk :studio_microphone:  

> Let us go back to 1990s (Kevin's Journey begins)
- Tie Fighter Wars (Motivation for Kevin to become a Video Game Dev).
- He ended up making websites though.
- Started with Developer Advocate Job and creating content for community.
- Started with TwilioQuest.

> [TwilioQuest](https://www.twilio.com/quest)
- Learn to code and lead your intrepid crew on a mission to save The Cloud in TwilioQuest, a PC role-playing game inspired by classics of the 16-bit era. Free forever, and available now for Windows, Mac, and Linux.

> The Game Design Stack
- Story
  - The narrative the player experiences - or better, controls.
  - Eg: Transistors
- World
  - The space, history, culture and constraints of the environment the player inhabits.
  - Eg: Breath Of Life
- Economy
  - How resources are created and exchanged inside the game.
  - Eg: Farm Simulator
- Action
  - What players will do in the game.
  - Eg: Sports Game
- Fantasy
  - Who or what the player will imagine themselves to be. How you will use empathy. 
  - Eg: Spider-man game on PS4

Explore above areas and come up with your own video game! :smiley: 

> Why it's a STACK ?

![stack](https://user-images.githubusercontent.com/41017419/101668752-f40d5b00-3a76-11eb-8637-a8ac96ef3e45.png)

> Important Points for TwilioQuest
- Content
- Authoring
- PC platform
- Narrative + level design

> Less Important Points for TwilioQuest
- World building, exposition
- XP and loot economy
- RPG staples

Know where is your game anchored from Game Design Stack and work on the experience!

> TwilioQuest Tech Stack : [Reference](https://www.twilio.com/blog/operators-manual-1-welcome-to-the-twilioquest-program)
- [Electron](https://electronjs.org/)
A cross-platform framework for building desktop apps with web technologies – used in the likes of Visual Studio Code and Slack. It was a natural fit for a body of contributors that are mostly web developers, and for an application that will be heavily dependent on displaying structured text and rich media (a major reason for not using something like Unity to build TwilioQuest). It also enables us to use the huge universe of packages available to Node.js applications.
- [React](https://reactjs.org/)
UI-heavy games like TwilioQuest benefit from having a robust rich client application framework like React.
- [Phaser](https://phaser.io/)
A mature framework for building 2D games that run in the browser. TwilioQuest’s graphical requirements are not intense, but thus far I’ve been quite pleased with the game’s performance on the WebGL canvas in Electron’s Chromium browser. Join Twilio in supporting Phaser on Patreon here!
- [Tiled](https://www.mapeditor.org/)
An open source map editor, this is the key tool we use to design environments in the next version of TwilioQuest.

- - -
If you have any questions, reach out to **[@kevinwhinnery](https://twitter.com/kevinwhinnery)**

> Do add your notes as well :smile:

**Reply by eddiejaoude** [reply]
These are EPIC!! thank you for sharing 👍

**Comment by siddharthdayalwal** [comment]
Will be sharing notes from **[GitHub Univers(ity) 2020 - Day 2](https://www.twitch.tv/githubeducation)**  🎉

# **`The university lab: experiment with the pack`** :package:  
> Get all [GitHub Education Benefits](https://education.github.com/)
- - -

## [Introduction to MongoDB and the GitHub Student Developer Program](https://www.twitch.tv/videos/831625369) :leaves:   
**Michael Lynn - Principal Developer Advocate, MongoDB**
**Lieke Boon - Program Manager, MongoDB for Academia**
> MongoDB is a document database with the scalability and flexibility that developers want and the querying and indexing that they need. In this talk, Michael and Lieke will provide an overview and demo of MongoDB and MongoDB Atlas, and talk about how developers can begin to incorporate these into their application stack.

Gist of the talk :studio_microphone:  

> [MongoDB is part of the GitHub Student Developer Pack](https://www.mongodb.com/students)

> [MongoDB for Academia](https://developer.mongodb.com/academia/)
- Emerging Developer Relations
- Create best learning experience for students
- Make resource for making student's career 

> [MongoDB for Educators](http://educators.mongodb.com/)
- Content support
- MongoDB University
- Academia Community

> [MongoDB for Students](http://students.mongodb.com/)
- Compaass
- MongoDB University
- MongoDB Atlas

> Future
- MongoDB Student Spotlights will be launching beginning 2021
- New course content for Educators will be published this quarter.

- - -
> [MongoDB Podcast](https://mongodb.com/podcast)

> [MongoDB Community](https://community.mongodb.com/)

> What is MongoDB?
- Document oriented
- Developer Friendly
- Highly scalable and available

> Comparing RDBMS with MongoDB - Terminology
- Collection of Row and Table(Data) is called Documents

> [MongoDB Atlas](https://www.mongodb.com/cloud/atlas)
- Move faster with a true multi-cloud database service for MongoDB built for agile teams who’d rather spend time building apps than managing databases.
- Available on AWS, Google Cloud, and Azure.
- Free Forever - M0
- Flexible options and pricing for larger, more powerful instances
- Auto-scale between instance sizes to simply handling unknown working development
- Builtin security in mind

> Demo [[Find the demo in recorded stream](https://www.twitch.tv/videos/831625369?t=16m04s)]
- Launch Cluster
- Deploy Sample Data
- Database Access - Add User
- Network Access - Add Firewall Entry
- Compass - Connect to Database

> Video Resources:
- [MongoDB Atlas Search - Product Demo](https://www.youtube.com/watch?v=kZ77X67GUfk)
- [An explanation of MongoDB Atlas' features and functionalities](https://www.youtube.com/watch?v=zbmMsLwMd-k)
- [Getting Your Free MongoDB Atlas Cluster](https://www.youtube.com/watch?v=rPqRyYJmx2g)

Find more videos and tutorials from [MongoDB YouTube Channel](https://www.youtube.com/user/MongoDB)

- - -
If you have any questions, reach out to **[@mlynn](https://twitter.com/mlynn)** and **[@Lieke2208](https://twitter.com/Lieke2208)**

> Do add your notes as well :smile:

**Comment by siddharthdayalwal** [comment]
Will be sharing notes from **[GitHub Univers(ity) 2020 - Day 2](https://www.twitch.tv/githubeducation)**  🎉

# **`The university lab: experiment with the pack`** :package:  
> Get all [GitHub Education Benefits](https://education.github.com/)
- - -

## [Using GitHub for multi-college collaborative undergraduate research](https://www.twitch.tv/videos/831625366) :mortar_board:    
**Dr. William "Chris" Headley - Associate Director, Electronic Systems Lab**
**Amos Johnson - Associate Professor of Computer Science, Morehouse College**
> Virginia Tech and Morehouse College are currently in the second semester of a Vertically Integrated Project (VIP). This experiential learning effort involves creating a long-term, multi-college, and multidisciplinary undergraduate research project in the technical area of deep learning applied to wireless communications. Due to COVID-19, the project has morphed into a completely online experience, with GitHub as the communications focal point. This change in structure has provided unique challenges in ensuring continued student engagement, collaboration, and research output—along with some unintended benefits. In this presentation, you’ll get an overview of the program, goals and motivations, and current status. In addition, the speakers will provide a unique look at how COVID-19 has led to necessary improvements in collaboration as a beneficial long term side-effect. You’ll come away with best practices for engaging students from diverse colleges, disciplines, and backgrounds effectively while maximizing student engagement and outcomes, especially in pure, or hybrid, online settings.

Gist of the talk :studio_microphone:  

> Project Background and Motivation 
- Virtual VIP Team Leaders
  - The Virginia Tech Hume C enter for National Security and technology has been a leader in Radio Frequency Machine Learning (RFML) research in recent years
  - Morehouse College for over 150 years have prepared young men to change the world through ethical leadership, driven by compassion, civility and integrity.
- Institutional & Technical Collaboration Goals
- VIP Collaboration Goals
- Diversification Goals

> Technical Overview
- **You Only Look Once Algorithm**
YOLO algorithm is an algorithm based on regression, instead of selecting the interesting part of an Image, it predicts classes and bounding boxes for the whole image in one run of the Algorithm. To understand the YOLO algorithm, first we need to understand what is actually being predicted.

## Spring 2020 Semester - Before Covid19

> Challenges
- Brand new project team with minimal exposure to wireless communication, spectrum sensing and machine learning.
- Diverse background of STEM and non-STEM majors with diff degrees of coding experience, necessitated increased collaboration and reference resource needs
- To facilitate inter college collaboration, each subteam had a mixture of VT and Morehouse students. 

> Spring 2020 Project Structure

![image](https://user-images.githubusercontent.com/41017419/101678859-8a944900-3a84-11eb-90b2-8a3ecee12440.png)

> Engagement Outcomes

![image](https://user-images.githubusercontent.com/41017419/101678922-a26bcd00-3a84-11eb-8bea-31be116d500a.png)

## Fall 2020 Semester - During Covid19

> Challenges
- Covid19 university restriction meant that this semester would be completely virtual, increasing the need for an improved project structure to facilitate remove collaboration
- Diverse background of STEM and non-STEM majors, with diff degrees of coding experience, necessitated increased collaboration and reference resource needs
- Increased from 2 technical subteams to 4 required more inter subteam coordinator resources

> Fall 2020 Project Structure

![image](https://user-images.githubusercontent.com/41017419/101679407-58371b80-3a85-11eb-9b0f-569e07dd7b51.png)

> Collaboration Outcomes

![image](https://user-images.githubusercontent.com/41017419/101679903-f88d4000-3a85-11eb-9b88-fbb714746cf7.png)

> Technical Outcomes
- Observation - Overall, students felt they learnt a lot about the technical area they were directly involved in but less so in areas where they were not actively contributing.

- Action - Need to provide students with more exposure to all aspects of the technical work for better holistic understanding. 

> GitHub Outcomes
- "I though [GitHub] was a very effective way to show our progress and what we're working towards"
- [GitHub Issues] kept me with things to do. It was great seeing deadlines and what was assigned."
- I have never used GitHub Issues, but I really like it so far. It see,s like a very effective way to track tasks and keep people accountable"

- - -
If you have any questions, reach out to **[Dr. William "Chris" Headley](https://www.linkedin.com/in/wchrisheadley/)** and **[@amosjohnsonjr](https://twitter.com/amosjohnsonjr)**

> Do add your notes as well :smile:

**Comment by siddharthdayalwal** [comment]
Will be sharing notes from **[GitHub Univers(ity) 2020 - Day 2](https://www.twitch.tv/githubeducation)**  🎉

# **`The university lab: experiment with the pack`** :package:  
> Get all [GitHub Education Benefits](https://education.github.com/)
- - -

## [Getting started in security research](https://www.twitch.tv/videos/831625367) :closed_lock_with_key:     
**Kevin Backhouse - Security Researcher, GitHub Security Lab**
> Computer security is an unusual field. Many of the most accomplished security researchers are either self-taught or did not come from a formal computer science background. For example, two of the most talented security researchers Kevin knows studied journalism in college. His own path into computer security was more conventional: he studied mathematics and computer science and then worked as a developer for 15 years. But he never expected to become a professional hacker— and his career transition into security was mostly an accident. So how do you get started in security research? There are many paths in, and security is a big field. In this talk, Kevin will give you the specific point of view of a former developer, and will focus on the secure code field. He will cover some of the techniques that he uses: code auditing, CodeQL, and fuzzing. He will also briefly discuss the different mindsets of a security researcher, compared to a developer.

Gist of the talk :studio_microphone:  

> [Common Vulnerabilities and Exposures](https://cve.mitre.org/) (CVE)
The Common Vulnerabilities and Exposures (CVE) system provides a reference-method for publicly known information-security vulnerabilities and exposures. ... The Security Content Automation Protocol uses CVE, and CVE IDs are listed on MITRE's system as well as in the US National Vulnerability Database.

> Venn Diagram 
![image](https://user-images.githubusercontent.com/41017419/101681494-37bc9080-3a88-11eb-8669-2b18119a9d06.png)

> Code Repositories

- Find codes at [github/securitylab](https://github.com/github/securitylab)

> Security Field as a whole

![image](https://user-images.githubusercontent.com/41017419/101682054-1b6d2380-3a89-11eb-9b9a-8fa2b2806c1d.png)

> Techniques
- Static analysis
- Manual audit
- Fuzzing
  - Hammer an application with ransom input until it crashed
  - Particularly effective on dense file formats like png/jpg/...
  - Easy to [get started with AFL](https://cambridge-community.org.uk/professional-development/gswafl/index.html)
    - [How to escape from the fuzz](http://securitylab.github.com/research/how-to-escape-from-the-fuzz)

> Difference in mindset of security researcher and developer

![image](https://user-images.githubusercontent.com/41017419/101682836-2aa0a100-3a8a-11eb-9c4a-92b083d394f2.png)

**"You're only as good as your last CVE"**

> Getting started
- Choose a project to focus on
- Study how it works
- Use all the tools
  - Static analysis
  - Manual audit
  - Fuzzing

> Reporting a bug
- Private Email, if possible
- Be polite
- Clear reproduction steps (PoC)
- Explain the bug
- Explain why its a security issue
- Disclosure Policy

> Hacker Glossary 

![image](https://user-images.githubusercontent.com/41017419/101683201-a7337f80-3a8a-11eb-9704-ced41fae2654.png)

- - -
If you have any questions, reach out to **[@kevin_backhouse](https://twitter.com/kevin_backhouse)**

> Do add your notes as well :smile:

**Reply by Vyvy-vi** [reply]
These are great notes @siddharthdayalwal  :)
🧡

**Reply by siddharthdayalwal** [reply]
Thanks @Vyvy-vi :smile:
